core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1721,8 +1720,10 @@ class ReplicaManager(val config: KafkaConfig,
       } else {
         // we do not need to check if the leader exists again since this has been done at the beginning of this process
         val partitionsToMakeFollowerWithLeaderAndOffset = partitionsToMakeFollower.map { partition =>
-          val leader = metadataCache.getAliveBrokers.find(_.id == partition.leaderReplicaIdOpt.get).get
-            .brokerEndPoint(config.interBrokerListenerName)
+          val leaderNode = metadataCache.getAliveBrokerNode(
+            partition.leaderReplicaIdOpt.getOrElse(-1), config.interBrokerListenerName.value()).
```

Comment:
It's a little strange to assume `getAliveBrokerNode` works with -1 as a broker id. Are we assuming that the implementation returns `noNode` in that case? I would probably try to write this like this:
```scala
partition.leaderReplicaIdOpt.flatMap(leaderId => 
  metadataCache.getAliveBrokerNode(leaderId, config.interBrokerListenerName.value))
 .getOrElse(Node.noNode())
```

Reply:
`getAliveBrokerNode` returns `None` when a broker ID that is not alive is passed. The conversion to `NoNode` happens in `ReplicaManager`, not in the metadata cache. I guess I don't feel strongly about it, though, if you want a different version of this...?